service/matching/task_queue_partition_manager.go

```diff
@@ -0,0 +1,707 @@
+	if err != nil {
+		return false, err
+	}
+
+	if pm.defaultQueue != dbq {
+		// default queue should stay alive even if requests go to other queues
+		pm.defaultQueue.MarkAlive()
+	}
+	return dbq.AddTask(ctx, params)
+}
+
+func (pm *taskQueuePartitionManagerImpl) PollTask(
+	ctx context.Context,
+	pollMetadata *pollMetadata,
+) (*internalTask, error) {
+	dbq := pm.defaultQueue
+	if pollMetadata.workerVersionCapabilities.GetUseVersioning() {
+		// default queue should stay alive even if requests go to other queues
+		pm.defaultQueue.MarkAlive()
+		// We don't need the userDataChanged channel here because:
+		// - if we sync match, we're done
+		// - if we spool to db, we'll re-resolve when it comes out of the db
+		userData, _, err := pm.GetUserData()
+		if err != nil {
+			return nil, err
+		}
+		dbq, err = pm.redirectToVersionSetQueueForPoll(ctx, pollMetadata.workerVersionCapabilities, userData)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if identity, ok := ctx.Value(identityKey).(string); ok && identity != "" {
+		pm.defaultQueue.UpdatePollerInfo(pollerIdentity(identity), pollMetadata)
+		// update timestamp when long poll ends
+		defer pm.defaultQueue.UpdatePollerInfo(pollerIdentity(identity), pollMetadata)
+	}
+	return dbq.PollTask(ctx, pollMetadata)
+}
+
+func (pm *taskQueuePartitionManagerImpl) DispatchSpooledTask(
+	ctx context.Context,
+	task *internalTask,
+) error {
+	taskInfo := task.event.GetData()
+	// This task came from taskReader so task.event is always set here.
+	directive := taskInfo.GetVersionDirective()
+	// Redirect and re-resolve if we're blocked in matcher and user data changes.
+	for {
+		dbq := pm.defaultQueue
+		var userDataChanged <-chan struct{}
+		if directive.GetValue() != nil {
+			userData, userDataCh, err := pm.GetUserData()
+			userDataChanged = userDataCh
+			if err != nil {
+				return err
+			}
+			dbq, err = pm.redirectToVersionSetQueueForAdd(ctx, directive, userData)
+			if err != nil {
+				return err
+			}
+		}
+		err := dbq.DispatchSpooledTask(ctx, task, userDataChanged)
+		if err != errInterrupted { // nolint:goerr113
+			return err
+		}
+	}
+}
+
+func (pm *taskQueuePartitionManagerImpl) DispatchQueryTask(
+	ctx context.Context,
+	taskID string,
+	request *matchingservice.QueryWorkflowRequest,
+) (*matchingservice.QueryWorkflowResponse, error) {
+	dbq, err := pm.getDbQueueForAdd(ctx, request.VersionDirective)
+	if err != nil {
+		return nil, err
+	}
+
+	if pm.defaultQueue != dbq {
+		// default queue should stay alive even if requests go to other queues
+		pm.defaultQueue.MarkAlive()
+	}
+
+	return dbq.DispatchQueryTask(ctx, taskID, request)
+}
+
+// GetUserData returns the user data for the task queue if any.
+// Note: can return nil value with no error.
+func (pm *taskQueuePartitionManagerImpl) GetUserData() (*persistencespb.VersionedTaskQueueUserData, <-chan struct{}, error) {
+	// mark alive so that it doesn't unload while a child partition is doing a long poll
+	pm.defaultQueue.MarkAlive()
```

Comment:
ah, I see what you did. this doesn't work because fetchUserData calls this, so it'll keep itself alive forever (I made the same mistake at first). it has to be called explicitly from matchingEngine. or I guess you could call pm.db.GetUserData from there? but I'm not sure I like that better